### PR TITLE
S3 writer should use multi-part upload.

### PR DESCRIPTION
**Pull request by dmvk**

### CODE REVIEW

File: direct/io-s3/src/main/java/cz/o2/proxima/direct/s3/S3Client.java [hunk 1]
```diff
@@ -45,6 +53,9 @@
 @Slf4j
 class S3Client implements Serializable {
 
+  /** Part size in multi-part upload (5MB). */
+  private static final int UPLOAD_PART_SIZE = 5 * 1024 * 1024;
```
Comment (je-ik) [inline direct/io-s3/src/main/java/cz/o2/proxima/direct/s3/S3Client.java:57]: Should this be configurable?

File: direct/io-s3/src/main/java/cz/o2/proxima/direct/s3/S3Client.java [hunk 2]
```diff
@@ -167,14 +176,82 @@ AmazonS3 client() {
...
+      public void close() throws IOException {
+        if (!closed) {
+          flush();
+          client()
+              .completeMultipartUpload(
+                  new CompleteMultipartUploadRequest(currentBucket, blobName, uploadId, eTags));
+          closed = true;
+        }
+      }
+    };
```
Comment (je-ik) [inline direct/io-s3/src/main/java/cz/o2/proxima/direct/s3/S3Client.java:255]: :+1:

File: direct/io-blob/src/test/java/cz/o2/proxima/direct/blob/BlobLogObservableTest.java [hunk 3]
```diff
@@ -105,6 +106,65 @@ public void onCompleted() {
...
+          }
+
+          @Override
+          public void onCompleted() {
+            // Noop.
+          }
+        });
+    assertTrue(errorReceived.await(5, TimeUnit.SECONDS));
+  }
+
```
[the remark this answers is not in the document]
Comment (dmvk) [reply direct/io-blob/src/test/java/cz/o2/proxima/direct/blob/BlobLogObservableTest.java:167]: yeah, the reader itself was super hard to mock in current design, so I've modelled the scenario using observer as they share the same codepath

File: direct/io-s3/src/main/java/cz/o2/proxima/direct/s3/S3BlobPath.java [hunk 4]
```diff
@@ -50,180 +38,51 @@
...
     }
 
     @Override
     public long getSize() {
-      return Optional.ofNullable(remoteObject)
-          .map(o -> o.getObjectMetadata().getContentLength())
-          .orElse(0L);
+      try (final S3Object object = fs.getObject(name)) {
+        return object.getObjectMetadata().getContentLength();
+      } catch (IOException e) {
```
[the remark this answers is not in the document]
Comment (dmvk) [reply direct/io-s3/src/main/java/cz/o2/proxima/direct/s3/S3BlobPath.java:56]: 👍

File: direct/io-s3/src/main/java/cz/o2/proxima/direct/s3/S3Client.java (hunk 1 above)
Comment (dmvk) [reply direct/io-s3/src/main/java/cz/o2/proxima/direct/s3/S3Client.java:57]: not in this iteration, this is a reasonable default (the same is used by native s3 client) and I don't think it needs to be override in most real world scenarios. We can always change this once we have a use case.

**Comment by je-ik** [comment]
I think it should be easy to get the coverage above 80%.

**Pull request merged**